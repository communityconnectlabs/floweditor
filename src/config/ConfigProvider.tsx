/* eslint-disable @typescript-eslint/no-explicit-any */
import { FlowEditorConfig } from 'flowTypes';
import * as React from 'react';

export const fakePropType: any = (): any => null;
fakePropType.isRequired = (): any => null;

interface ConfigProviderProps {
  config: FlowEditorConfig;
  children: any;
}

export interface ConfigProviderContext {
  config: FlowEditorConfig;
}

// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR = 'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
  'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<ConfigProviderProps> {
  public static childContextTypes = {
    config: fakePropType,
    assetService: fakePropType
  };

  constructor(props: ConfigProviderProps) {
    super(props);
    const win = window as any;
    if (win.isMobile && win.isMobile()) {
      this.props.config.mutable = false;
    }

    if (React.Children.count(props.children) > 1) {
      throw new Error(SINGLE_CHILD_ERROR);
    } else if (!React.isValidElement(props.children)) {
      throw new Error(VALID_CHILD_ERROR);
    }
  }

  public getChildContext(): ConfigProviderContext {
    return {
      config: this.props.config
    };
  }

  public render(): JSX.Element {
    return React.Children.only(this.props.children);
  }
}
