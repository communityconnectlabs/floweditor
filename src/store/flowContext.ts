import { FlowDefinition, FlowNode, UINode } from 'flowTypes';
import { combineReducers } from 'redux';
import ActionTypes, {
  UpdateAssetsAction,
  UpdateBaseLanguageAction,
  UpdateContactFieldsAction,
  UpdateDefinitionAction,
  UpdateDependenciesAction,
  UpdateNodesAction
} from 'store/actionTypes';
import Constants from 'store/constants';

// tslint:disable:no-shadowed-variable
export interface RenderNodeMap {
  [uuid: string]: RenderNode;
}

export interface RenderNode {
  ui: UINode;
  node: FlowNode;
  inboundConnections: { [nodeUUID: string]: string };
  ghost?: boolean;
}

export interface FunctionExample {
  template: string;
  output: string;
}

export interface CompletionOption {
  name?: string;
  summary: string;

  // functions
  signature?: string;
  detail?: string;
  examples?: FunctionExample[];
}

export interface ContactFields {
  [snakedFieldName: string]: string;
}

export enum AssetType {
  Channel = 'channel',
  Classifier = 'classifier',
  Contact = 'contact',
  ContactProperty = 'property',
  Currency = 'currency',
  Environment = 'environment',
  Expression = 'expression',
  Field = 'field',
  Flow = 'flow',
  Global = 'global',
  Group = 'group',
  Label = 'label',
  Language = 'language',
  Remove = 'remove',
  Resthook = 'resthook',
  Result = 'result',
  Revision = 'revision',
  Scheme = 'scheme',
  Template = 'template',
  URN = 'urn'
}

export interface Reference {
  nodeUUID: string;
  actionUUID?: string;
}

export interface Asset {
  id: string;
  name: string;
  type: AssetType;

  order?: number;
  references?: Reference[];
  isNew?: boolean;
  content?: any;
}

export const REMOVE_VALUE_ASSET = {
  id: AssetType.Remove,
  name: 'Remove Value',
  type: AssetType.Remove
};

export const DEFAULT_LANGUAGE = {
  id: 'base',
  name: 'Default',
  type: AssetType.Language
};

export interface AssetStore {
  [assetType: string]: Assets;
}

export interface AssetMap {
  [id: string]: Asset;
}

export interface Assets {
  // our local cache of assets
  items: AssetMap;

  type: AssetType;

  // an optional endpoint to search for more
  endpoint?: string;

  // option name for the id when fetching remotely
  id?: string;

  // have our assets already been loaded
  prefetched?: boolean;
}

export interface FlowContext {
  dependencies: FlowDefinition[];
  baseLanguage: Asset;
  contactFields: ContactFields;
  definition: FlowDefinition;
  nodes: { [uuid: string]: RenderNode };
  assetStore: AssetStore;
}

// Initial state
export const initialState: FlowContext = {
  definition: null,
  dependencies: null,
  baseLanguage: null,
  contactFields: {},
  nodes: {},
  assetStore: {}
};

// Action Creators
export const updateDefinition = (definition: FlowDefinition): UpdateDefinitionAction => ({
  type: Constants.UPDATE_DEFINITION,
  payload: {
    definition
  }
});

export const updateNodes = (nodes: RenderNodeMap): UpdateNodesAction => ({
  type: Constants.UPDATE_NODES,
  payload: {
    nodes
  }
});

export const updateDependencies = (dependencies: FlowDefinition[]): UpdateDependenciesAction => ({
  type: Constants.UPDATE_DEPENDENCIES,
  payload: {
    dependencies
  }
});

export const updateBaseLanguage = (baseLanguage: Asset): UpdateBaseLanguageAction => ({
  type: Constants.UPDATE_BASE_LANGUAGE,
  payload: {
    baseLanguage
  }
});

export const updateContactFields = (contactFields: ContactFields): UpdateContactFieldsAction => ({
  type: Constants.UPDATE_CONTACT_FIELDS,
  payload: {
    contactFields
  }
});

export const updateAssets = (assets: AssetStore): UpdateAssetsAction => ({
  type: Constants.UPDATE_ASSET_MAP,
  payload: {
    assets
  }
});

// Reducers
export const definition = (
  state: FlowDefinition = initialState.definition,
  action: ActionTypes
) => {
  switch (action.type) {
    case Constants.UPDATE_DEFINITION:
      return action.payload.definition;
    default:
      return state;
  }
};

export const nodes = (state: {} = initialState.nodes, action: ActionTypes) => {
  switch (action.type) {
    case Constants.UPDATE_NODES:
      return action.payload.nodes;
    default:
      return state;
  }
};

export const dependencies = (
  state: FlowDefinition[] = initialState.dependencies,
  action: ActionTypes
) => {
  switch (action.type) {
    case Constants.UPDATE_DEPENDENCIES:
      return action.payload.dependencies;
    default:
      return state;
  }
};

export const assetStore = (state: AssetStore = initialState.assetStore, action: ActionTypes) => {
  switch (action.type) {
    case Constants.UPDATE_ASSET_MAP:
      return action.payload.assets;
    default:
      return state;
  }
};

export const baseLanguage = (state: Asset = initialState.baseLanguage, action: ActionTypes) => {
  switch (action.type) {
    case Constants.UPDATE_BASE_LANGUAGE:
      return action.payload.baseLanguage;
    default:
      return state;
  }
};

export const contactFields = (
  state: ContactFields = initialState.contactFields,
  action: ActionTypes
) => {
  switch (action.type) {
    case Constants.UPDATE_CONTACT_FIELDS:
      return action.payload.contactFields;
    default:
      return state;
  }
};

// Root reducer
export default combineReducers({
  definition,
  nodes,
  dependencies,
  assetStore,
  baseLanguage,
  contactFields
});
