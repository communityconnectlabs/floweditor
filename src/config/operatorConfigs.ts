import { HIDDEN, Operator, OperatorMap, Operators, TEXT_TYPES, ONLINE } from 'config/interfaces';

export const intentOperatorList: Operator[] = [
  {
    type: Operators.has_top_intent,
    verboseName: 'has top intent',
    operands: 2,
    visibility: ONLINE
  },
  {
    type: Operators.has_intent,
    verboseName: 'has intent',
    operands: 2,
    visibility: ONLINE
  }
];

export const operatorConfigList: Operator[] = [
  {
    type: Operators.has_any_word,
    verboseName: 'has any of the words',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_all_words,
    verboseName: 'has all of the words',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_phrase,
    verboseName: 'has the phrase',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_only_phrase,
    verboseName: 'has only the phrase',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_beginning,
    verboseName: 'starts with',
    operands: 1
  },
  {
    type: Operators.has_text,
    verboseName: 'has some text',
    operands: 0,
    categoryName: 'Has Text',
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_number,
    verboseName: 'has a number',
    operands: 0,
    categoryName: 'Has Number'
  },
  {
    type: Operators.has_number_between,
    verboseName: 'has a number between',
    operands: 2
  },
  {
    type: Operators.has_number_lt,
    verboseName: 'has a number below',
    operands: 1
  },
  {
    type: Operators.has_number_lte,
    verboseName: 'has a number at or below',
    operands: 1
  },
  {
    type: Operators.has_number_eq,
    verboseName: 'has a number equal to',
    operands: 1
  },
  {
    type: Operators.has_number_gte,
    verboseName: 'has a number at or above',
    operands: 1
  },
  {
    type: Operators.has_number_gt,
    verboseName: 'has a number above',
    operands: 1
  },
  {
    type: Operators.has_date,
    verboseName: 'has a date',
    operands: 0,
    categoryName: 'Has Date',
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_date_lt,
    verboseName: 'has a date before',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_date_eq,
    verboseName: 'has a date equal to',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_date_gt,
    verboseName: 'has a date after',
    operands: 1,
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_time,
    verboseName: 'has a time',
    operands: 0,
    categoryName: 'Has Time',
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_group,
    verboseName: 'is in the group',
    operands: 1,
    visibility: HIDDEN
  },
  {
    type: Operators.has_category,
    verboseName: 'has the category',
    operands: 0,
    visibility: HIDDEN
  },
  {
    type: Operators.has_phone,
    verboseName: 'has a phone number',
    operands: 0,
    categoryName: 'Has Phone'
  },
  {
    type: Operators.has_email,
    verboseName: 'has an email',
    operands: 0,
    categoryName: 'Has Email',
    visibility: TEXT_TYPES
  },
  {
    type: Operators.has_state,
    verboseName: 'has state',
    operands: 0,
    categoryName: 'Has State'
  },
  {
    type: Operators.has_district,
    verboseName: 'has district',
    operands: 1,
    categoryName: 'Has District'
  },
  {
    type: Operators.has_ward,
    verboseName: 'has ward',
    operands: 2,
    categoryName: 'Has Ward'
  },
  {
    type: Operators.has_error,
    verboseName: 'has an error',
    operands: 0,
    categoryName: 'Has Error',
    visibility: HIDDEN
  },
  {
    type: Operators.has_value,
    verboseName: 'is not empty',
    operands: 0,
    categoryName: 'Not Empty'
  },
  {
    type: Operators.has_pattern,
    verboseName: 'matches regex',
    operands: 1
  }
];

export const operatorConfigMap: OperatorMap = [...operatorConfigList, ...intentOperatorList].reduce(
  (map: OperatorMap, operatorConfig: Operator) => {
    map[operatorConfig.type] = operatorConfig;
    return map;
  },
  {}
);

/**
 * Shortcut for constant lookup of operator config in operator configs map
 * @param {string} type - The type of the operator config to return, e.g. 'send_msg'
 * @returns {Object} - The operator config found at operatorConfigs[type] or -1
 */
export const getOperatorConfig = (type: Operators): Operator => operatorConfigMap[type];
