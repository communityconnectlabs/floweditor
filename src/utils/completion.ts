import { AssetStore, CompletionOption } from 'store/flowContext';

export interface CompletionAssets {
  assetStore: AssetStore;
  schema: CompletionSchema;
}

export interface CompletionProperty {
  key: string;
  help: string;
  type: string;
}

export interface CompletionType {
  name: string;

  key_source?: string;
  property_template?: CompletionProperty;
  properties?: CompletionProperty[];
}

export interface CompletionSchema {
  types: CompletionType[];
  root: CompletionProperty[];
}

export const getFunctions = (functions: CompletionOption[], query: string): CompletionOption[] => {
  return functions.filter((option: CompletionOption) => {
    if (option.signature) {
      return option.signature.indexOf(query) === 0;
    }
    return false;
  });
};

/**
 * Takes a dot query and returns the completions options at the current level
 * @param dotQuery query such as "results.field_name.ca"
 */
export const getCompletions = (
  completion: CompletionAssets,
  dotQuery: string
): CompletionOption[] => {
  const parts = (dotQuery || '').split('.');
  let currentProps: CompletionProperty[] = completion.schema.root;

  let prefix = '';
  let part = '';
  while (parts.length > 0) {
    part = parts.shift();
    if (part) {
      // eslint-disable-next-line
      const nextProp = currentProps.find((prop: CompletionProperty) => prop.key === part);
      if (nextProp) {
        // eslint-disable-next-line
        const nextType = completion.schema.types.find(
          (type: CompletionType) => type.name === nextProp.type
        );
        if (nextType && nextType.properties) {
          currentProps = nextType.properties;
          prefix += part + '.';
        } else if (nextType && nextType.property_template) {
          prefix += part + '.';
          const template = nextType.property_template;
          currentProps = Object.keys(completion.assetStore[nextType.name].items).map(
            (key: string) => ({
              key: template.key.replace('{key}', key),
              help: template.help.replace('{key}', key),
              type: template.type
            })
          );
        } else {
          // eslint-disable-next-line
          currentProps = currentProps.filter((prop: CompletionProperty) =>
            prop.key.startsWith(part.toLowerCase())
          );
          break;
        }
      } else {
        // eslint-disable-next-line
        currentProps = currentProps.filter((prop: CompletionProperty) =>
          prop.key.startsWith(part.toLowerCase())
        );
        break;
      }
    }
  }

  return currentProps.map((prop: CompletionProperty) => {
    const name =
      prop.key === '__default__' ? prefix.substr(0, prefix.length - 1) : prefix + prop.key;
    return { name, summary: prop.help };
  });
};

export const getCompletionName = (option: CompletionOption): string => {
  return option.name || option.signature.substr(0, option.signature.indexOf('('));
};

export const getCompletionSignature = (option: CompletionOption): string => {
  return option.signature.substr(option.signature.indexOf('('));
};
