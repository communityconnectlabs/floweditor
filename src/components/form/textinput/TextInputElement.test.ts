import { KeyValues } from 'components/form/textinput/constants';
import {
  TextInputElement,
  TextInputProps,
  TextInputState
} from 'components/form/textinput/TextInputElement';
import { Types } from 'config/interfaces';
import { getTypeConfig } from 'config/typeConfigs';
import setCaretPosition from 'get-input-selection';
import { AssetType } from 'store/flowContext';
import * as completionSchema from 'test/assets/completion.json';
import functions from 'test/assets/functions.json';
import { composeComponentTestUtils } from 'testUtils';

// we need to track where our cursor would be to simulate properly
let mockCursor = 0;

const baseProps: TextInputProps = {
  name: 'Message',
  typeConfig: getTypeConfig(Types.send_msg),
  assetStore: { fields: { items: {}, type: AssetType.Field } },
  completionSchema,
  functions
};

const { setup, spyOn } = composeComponentTestUtils(TextInputElement, baseProps);

jest.mock('get-input-selection', () => ({
  default: jest.fn()
}));

const mockEvent = {
  preventDefault(): void {
    return;
  },
  stopPropagation(): void {
    return;
  }
};

const simulateString = (wrap: any, keys: string) => {
  const input = wrap.find('textarea');

  const value: string = wrap.instance().textEl.value;
  let result = value.substr(0, mockCursor);
  const end = value.substr(mockCursor);

  for (const key of keys) {
    input.prop('onKeyDown')({
      ...mockEvent,
      key
    });

    result += key;
    mockCursor++;

    input.prop('onKeyUp')({
      ...mockEvent,
      key
    });

    input.prop('onChange')({
      currentTarget: {
        value: result + end,
        selectionStart: mockCursor
      }
    });
  }
};

const simulateKey = (wrap: any, key: KeyValues, ctrlKey: boolean = false, presses: number = 1) => {
  for (let i = 0; i < presses; i++) {
    if (key === KeyValues.KEY_AT) {
      simulateString(wrap, key);
    } else {
      const input = wrap.find('textarea');
      const startLength = wrap.instance().textEl.value.length;

      // fire a key event
      input.prop('onKeyDown')({
        ...mockEvent,
        key,
        ctrlKey,
        currentTarget: {
          value: wrap.instance().textEl.value,
          selectionStart: mockCursor
        }
      });

      // fire a key event
      input.prop('onKeyUp')({
        ...mockEvent,
        key,
        ctrlKey,
        currentTarget: {
          value: wrap.instance().textEl.value,
          selectionStart: mockCursor
        }
      });

      // update our mock cursor accordingly
      if (key === KeyValues.KEY_LEFT) {
        mockCursor--;
      } else if (key === KeyValues.KEY_RIGHT) {
        mockCursor++;
      } else {
        mockCursor += wrap.instance().textEl.value.length - startLength;
      }
    }
  }
};

const getState = (wrap: any): TextInputState => {
  return wrap.instance().state as TextInputState;
};

const createWrapper = () => {
  return setup(false, {
    $merge: {
      onChange: jest.fn(),
      textarea: true,
      autocomplete: true,
      onFieldFailures: jest.fn(),
      onBlur: jest.fn()
    }
  }).wrapper;
};

describe(TextInputElement.name, () => {
  beforeEach(() => {
    mockCursor = 0;
  });
  afterEach(() => {
    setCaretPosition.default.mockReset();
  });

  describe('function context', () => {
    it('should use the most recent incomplete function', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@(max(default(contact.rank, cont');

      const state = getState(wrapper);
      expect(state.completionVisible).toBeTruthy();
      expect(state.query).toEqual('cont');
      expect(state.matches.length).toBe(1);
      expect(state.matches).toMatchSnapshot();
      expect(state.fn.signature).toEqual('default(value, default)');
    });

    it('should use fold back on function completion', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@(max(default(contact.rank, contact.age), ');

      const state = getState(wrapper);
      expect(state.query).toEqual('');
      expect(state.fn.signature).toEqual('max(numbers...)');
    });
  });

  describe('filtering', () => {
    it('should bring up completion menu for top level options', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@');

      const state = getState(wrapper);
      expect(state.completionVisible).toBeTruthy();
      expect(state.query).toEqual('');

      // should show all top level options
      expect(state.matches).toMatchSnapshot();
    });

    it('should show filter options', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@co');

      const state = getState(wrapper);
      expect(state.completionVisible).toBeTruthy();
      expect(state.query).toEqual('co');

      // only our contact option should be there
      expect(state.matches.length).toBe(1);
      expect(state.matches).toMatchSnapshot();
    });

    it('should show filter at the second level', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@contact.first');

      const state = getState(wrapper);
      expect(state.completionVisible).toBeTruthy();
      expect(state.query).toEqual('contact.first');

      // should see first name
      expect(state.matches.length).toBe(1);
      expect(state.matches).toMatchSnapshot();
    });

    it('should not match functions without open paren', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@max');

      const state = getState(wrapper);
      expect(state.matches.length).toBe(0);
      expect(state.completionVisible).toBeFalsy();
    });

    it('should bring up completion menu for top level options and functions', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@(');

      const state = getState(wrapper);
      // should show all top level options and functions
      expect(state.matches.length).toBeGreaterThan(10);
      expect(state.completionVisible).toBeTruthy();
    });

    it('should bring up functions within text', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, 'some text @(');

      const state = getState(wrapper);
      expect(state.matches.length).toBeGreaterThan(10);
      expect(state.completionVisible).toBeTruthy();
    });
  });

  describe('completion', () => {
    it('should handle completion w/ "Enter" key', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@cont');
      simulateKey(wrapper, KeyValues.KEY_ENTER);

      const state = getState(wrapper);
      expect(state.completionVisible).toBeFalsy();
      expect(state.query).toEqual('');
      expect(state.matches.length).toBe(0);
    });

    it('should handle completion w/ "Tab" key', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@cont');
      simulateKey(wrapper, KeyValues.KEY_TAB);

      const state = getState(wrapper);
      expect(state.completionVisible).toBeTruthy();
      expect(state.query).toEqual('contact');

      // tabbing forward should give us all our contact options
      expect(state.matches).toMatchSnapshot();
    });

    it('should complete in the middle of text', () => {
      const wrapper = createWrapper();

      // type some text then move cursor back
      simulateString(wrapper, 'hello world');
      simulateKey(wrapper, KeyValues.KEY_LEFT, false, 6);

      // enter a query and complete it
      simulateString(wrapper, ' @cont');
      simulateKey(wrapper, KeyValues.KEY_TAB);

      const state = getState(wrapper);
      expect(state.completionVisible).toBeTruthy();
      expect(state.query).toEqual('contact');
      expect(state.value).toBe('hello @contact world');

      // tabbing forward should give us all our contact options
      expect(state.matches).toMatchSnapshot();
    });

    it('should complete over adjacent text', () => {
      const wrapper = createWrapper();

      // type some text then move cursor back
      simulateString(wrapper, 'hello contact');
      simulateKey(wrapper, KeyValues.KEY_LEFT, false, 7);
      simulateKey(wrapper, KeyValues.KEY_AT);
      simulateKey(wrapper, KeyValues.KEY_TAB);

      const state = getState(wrapper);
      expect(state.completionVisible).toBeTruthy();
      expect(state.value).toBe('hello @contact');
    });

    it('should complete over dot extensions', () => {
      const wrapper = createWrapper();

      // type some text then move cursor back
      simulateString(wrapper, 'hello @contact.first_name');
      simulateKey(wrapper, KeyValues.KEY_LEFT, false, 17);
      simulateString(wrapper, 'h');
      simulateKey(wrapper, KeyValues.KEY_TAB);

      const state = getState(wrapper);
      expect(state.completionVisible).toBeTruthy();
      expect(state.value).toBe('hello @child');
    });
  });

  describe('navigation', () => {
    it('should filter after tabbing', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@cont');
      simulateKey(wrapper, KeyValues.KEY_TAB);
      simulateString(wrapper, '.fir');

      const state = getState(wrapper);
      expect(state.completionVisible).toBeTruthy();
      expect(state.query).toEqual('contact.fir');

      // tabbing forward should give us all our contact options
      expect(state.matches.length).toBe(1);
      expect(state.matches[0].name).toBe('contact.first_name');
      expect(state.matches).toMatchSnapshot();
    });

    it('should allow navigation with arrow keys', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@c');

      simulateKey(wrapper, KeyValues.KEY_DOWN);
      let state = getState(wrapper);
      expect(state.selectedOptionIndex).toBe(1);

      simulateKey(wrapper, KeyValues.KEY_UP);
      state = getState(wrapper);
      expect(state.selectedOptionIndex).toBe(0);
    });

    it('should allow navigation with OSX shortcuts', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@c');

      simulateKey(wrapper, KeyValues.KEY_N, true);
      let state = getState(wrapper);
      expect(state.selectedOptionIndex).toBe(1);

      simulateKey(wrapper, KeyValues.KEY_P, true);
      state = getState(wrapper);
      expect(state.selectedOptionIndex).toBe(0);
    });
  });

  describe('visibility', () => {
    it('should check contact fields on blur', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@fields.missing');
      wrapper.instance().handleBlur({});
      expect(wrapper.props().onFieldFailures).toHaveBeenCalled();
    });

    it('should hide if outside completed expression', () => {
      const wrapper = createWrapper();
      simulateString(wrapper, '@(default(contact.first_name, contact.name))');

      const state = getState(wrapper);
      expect(state.completionVisible).toBeFalsy();
      expect(state.matches.length).toBe(0);
    });

    it('should hide if navigating out of expression', () => {
      const wrapper = createWrapper();

      // enter some text, backpedal and insert an expression
      simulateString(wrapper, 'hello world');
      simulateKey(wrapper, KeyValues.KEY_LEFT, false, 6);
      simulateString(wrapper, ' @cont');

      // now move right until we are out of the expression
      simulateKey(wrapper, KeyValues.KEY_RIGHT, false, 4);

      const state = getState(wrapper);
      expect(state.completionVisible).toBeFalsy();
      expect(state.matches.length).toBe(0);
    });

    it('should show completion with ctrl+space', () => {
      const wrapper = createWrapper();

      // enter some text, backpedal and insert an expression
      simulateString(wrapper, '@contact.first_name is ready');
      simulateKey(wrapper, KeyValues.KEY_LEFT, false, 10);

      // shouldn't see completion yet
      expect(getState(wrapper).completionVisible).toBeFalsy();

      // now do the Ctrl+Space shortcut to show completion
      simulateKey(wrapper, KeyValues.KEY_SPACE, true);
      expect(getState(wrapper).completionVisible).toBeTruthy();
    });
  });
});
