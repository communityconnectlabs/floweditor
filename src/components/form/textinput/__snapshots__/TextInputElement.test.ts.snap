// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputElement completion should complete in the middle of text 1`] = `
Array [
  Object {
    "name": "contact",
    "summary": "the name or URN",
  },
  Object {
    "name": "contact.uuid",
    "summary": "the UUID of the contact",
  },
  Object {
    "name": "contact.id",
    "summary": "the numeric ID of the contact",
  },
  Object {
    "name": "contact.first_name",
    "summary": "the first name of the contact",
  },
  Object {
    "name": "contact.name",
    "summary": "the name of the contact",
  },
  Object {
    "name": "contact.language",
    "summary": "the language of the contact as 3-letter ISO code",
  },
  Object {
    "name": "contact.created_on",
    "summary": "the creation date of the contact",
  },
  Object {
    "name": "contact.urns",
    "summary": "the URNs belonging to the contact",
  },
  Object {
    "name": "contact.urn",
    "summary": "the preferred URN of the contact",
  },
  Object {
    "name": "contact.groups",
    "summary": "the groups the contact belongs to",
  },
  Object {
    "name": "contact.fields",
    "summary": "the custom field values of the contact",
  },
  Object {
    "name": "contact.channel",
    "summary": "the preferred channel of the contact",
  },
]
`;

exports[`TextInputElement completion should handle completion w/ "Tab" key 1`] = `
Array [
  Object {
    "name": "contact",
    "summary": "the name or URN",
  },
  Object {
    "name": "contact.uuid",
    "summary": "the UUID of the contact",
  },
  Object {
    "name": "contact.id",
    "summary": "the numeric ID of the contact",
  },
  Object {
    "name": "contact.first_name",
    "summary": "the first name of the contact",
  },
  Object {
    "name": "contact.name",
    "summary": "the name of the contact",
  },
  Object {
    "name": "contact.language",
    "summary": "the language of the contact as 3-letter ISO code",
  },
  Object {
    "name": "contact.created_on",
    "summary": "the creation date of the contact",
  },
  Object {
    "name": "contact.urns",
    "summary": "the URNs belonging to the contact",
  },
  Object {
    "name": "contact.urn",
    "summary": "the preferred URN of the contact",
  },
  Object {
    "name": "contact.groups",
    "summary": "the groups the contact belongs to",
  },
  Object {
    "name": "contact.fields",
    "summary": "the custom field values of the contact",
  },
  Object {
    "name": "contact.channel",
    "summary": "the preferred channel of the contact",
  },
]
`;

exports[`TextInputElement filtering should bring up completion menu for top level options 1`] = `
Array [
  Object {
    "name": "contact",
    "summary": "the contact",
  },
  Object {
    "name": "fields",
    "summary": "the custom field values of the contact",
  },
  Object {
    "name": "urns",
    "summary": "the URN values of the contact",
  },
  Object {
    "name": "results",
    "summary": "the current run results",
  },
  Object {
    "name": "input",
    "summary": "the current input from the contact",
  },
  Object {
    "name": "run",
    "summary": "the current run",
  },
  Object {
    "name": "child",
    "summary": "the last child run",
  },
  Object {
    "name": "parent",
    "summary": "the parent of the run",
  },
  Object {
    "name": "webhook",
    "summary": "the parsed JSON response of the last webhook call",
  },
  Object {
    "name": "trigger",
    "summary": "the trigger that started this session",
  },
]
`;

exports[`TextInputElement filtering should show filter at the second level 1`] = `
Array [
  Object {
    "name": "contact.first_name",
    "summary": "the first name of the contact",
  },
]
`;

exports[`TextInputElement filtering should show filter options 1`] = `
Array [
  Object {
    "name": "contact",
    "summary": "the contact",
  },
]
`;

exports[`TextInputElement function context should use the most recent incomplete function 1`] = `
Array [
  Object {
    "name": "contact",
    "summary": "the contact",
  },
]
`;

exports[`TextInputElement navigation should filter after tabbing 1`] = `
Array [
  Object {
    "name": "contact.first_name",
    "summary": "the first name of the contact",
  },
]
`;
